core/src/impl/Kokkos_Core.cpp

```diff
@@ -281,57 +310,92 @@ int Kokkos::Impl::get_ctest_gpu(const char* local_rank_str) {
-      use_gpu = get_ctest_gpu(local_rank_str);
-    } else if (ndevices > 0) {
-      // Use the device assigned by the rank
-      if (local_rank_str) {
-        auto local_rank = std::stoi(local_rank_str);
-        use_gpu         = local_rank % ndevices;
-      } else {
-        // user only gave use ndevices, but the MPI environment variable wasn't
-        // set. start with GPU 0 at this point
-        use_gpu = 0;
+std::vector<int> Kokkos::Impl::get_visible_devices(
+    Kokkos::InitializationSettings const& settings, int device_count) {
+  std::vector<int> visible_devices;
+  char* env_visible_devices = std::getenv("KOKKOS_VISIBLE_DEVICES");
+  if (env_visible_devices) {
+    std::stringstream ss(env_visible_devices);
+    for (int i; ss >> i;) {
+      visible_devices.push_back(i);
+      if (ss.peek() == ',') ss.ignore();
+    }
+  } else {
+    int num_devices =
+        settings.has_num_devices() ? settings.get_num_devices() : device_count;
+    for (int i = 0; i < num_devices; ++i) {
+      visible_devices.push_back(i);
+    }
+    if (settings.has_skip_device()) {
+      if (visible_devices.size() == 1 && settings.get_skip_device() == 0) {
+        Kokkos::abort(
+            "Error: skipping the only GPU available for execution.\n");
       }
+      visible_devices.erase(
+          std::remove(visible_devices.begin(), visible_devices.end(),
+                      settings.get_skip_device()),
+          visible_devices.end());
     }
-    // shift assignments over by one so no one is assigned to "skip_device"
-    if (use_gpu >= skip_device) ++use_gpu;
   }
-  return use_gpu;
+  if (visible_devices.empty()) {
+    Kokkos::abort("Error: no GPU available for execution.\n");
+  }
+  return visible_devices;
+}
+
+int Kokkos::Impl::get_gpu(const InitializationSettings& settings) {
+  std::vector<int> visible_devices =
+      get_visible_devices(settings, get_device_count());
+  // device_id is provided
+  if (settings.has_device_id()) {
+    return visible_devices[settings.get_device_id()];
+  }
+  // by default use the first GPU available for execution
+  // (neither device_id nor map_device_id_by are provided)
+  if (!settings.has_map_device_id_by()) {
+    return visible_devices[0];
+  }
+  // map_device_id provided
+  // either random or round-robin assignement based on local MPI rank
+  if (!is_valid_map_device_id_by(settings.get_map_device_id_by())) {
+    std::cerr << "Warning: unrecognized map_device_id_by setting \""
+              << settings.get_map_device_id_by() << "\" ignored."
+              << " Raised by Kokkos::initialize(int argc, char* argv[])."
+              << std::endl;
+    return visible_devices[0];
+  }
+
+  if (settings.get_map_device_id_by() == "random") {
+    std::default_random_engine gen(get_process_id());
+    std::uniform_int_distribution<int> distribution(0,
+                                                    visible_devices.size() - 1);
+    return visible_devices[distribution(gen)];
+  }
```

Comment:
could we end up with multiple processes using the same gpu in this case? e.g. multiple kokkos processes getting the same device id randomly ... 
Is there a real use case for the random assignment?

Reply:
yes we can, but in the primary use case that is not unexpected anyway. The main use case for this is parallel CI. Essentially running ctest with -j. Most unit tests don't tax GPUs dramatically and are not timing sensitive. But it helps if they don't all hit the same GPU. In practice we can run for example most Trilinos testing with -j16 on a 4 GPU machine. But if all 16 hit GPU zero we run out of memory occasionally.

Reply:
The problem there is that ctest doesn't have resource management, and its not MPI i.e. the round robin doesn't work. Random assignement is good enough in spreading out however.

Reply:
thanks